aries_vcx/src/protocols/connection/mod.rs

```diff
@@ -1,3 +1,188 @@
+mod generic;
+pub mod initiation_type;
 pub mod invitee;
 pub mod inviter;
 pub mod pairwise_info;
+mod serializable;
+mod trait_bounds;
+
+use messages::{
+    a2a::{protocol_registry::ProtocolRegistry, A2AMessage},
+    diddoc::aries::diddoc::AriesDidDoc,
+    protocols::{
+        connection::problem_report::{ProblemCode, ProblemReport},
+        discovery::disclose::{Disclose, ProtocolDescriptor},
+    },
+};
+use std::{error::Error, sync::Arc};
+
+use crate::{
+    errors::error::{AriesVcxError, AriesVcxErrorKind, VcxResult},
+    plugins::wallet::base_wallet::BaseWallet,
+    transport::Transport,
+    utils::encryption_envelope::EncryptionEnvelope,
+};
+
+use self::{
+    generic::GenericState,
+    pairwise_info::PairwiseInfo,
+    trait_bounds::{CompleteState, HandleProblem, TheirDidDoc, ThreadId},
+};
+
+pub use self::generic::{GenericConnection, State, ThinState};
+
+/// A state machine for progressing through the (connection protocol)[https://github.com/hyperledger/aries-rfcs/blob/main/features/0160-connection-protocol/README.md].
+#[derive(Clone, Deserialize)]
+#[serde(try_from = "GenericConnection")]
+#[serde(bound = "(I, S): TryFrom<GenericState, Error = AriesVcxError>")]
+pub struct Connection<I, S> {
+    source_id: String,
+    pairwise_info: PairwiseInfo,
+    initiation_type: I,
+    state: S,
+}
+
+impl<I, S> Connection<I, S> {
+    pub(crate) fn from_parts(source_id: String, pairwise_info: PairwiseInfo, initiation_type: I, state: S) -> Self {
```

Comment:
Just leaving a comment here in-case we decide in tomorrow's meeting that we should make the method `pub`.